File: lib/src/index_map.dart
import 'dart:convert';

import 'tokenizer.dart';

typedef DocPropIndex = List<int>;

class IndexMap {
  final Map<String, List<DocPropIndex>> _data = {};
  final List<String> _keys = [];
  final List<String> fields;
  final List<String> _effectiveFields = [];
  final Tokenizer tokenizer;

  IndexMap({required this.tokenizer, this.fields = const []}) {
    if (fields.isNotEmpty) {
      _effectiveFields.addAll(fields);
    }
  }

  void add(String key, Map<String, dynamic> document) {
    _keys.add(key);
    for (var entry in document.entries) {
      if (entry.value is! String ||
          (fields.isNotEmpty && !fields.contains(entry.key))) {
        continue;
      }
      final tokens = tokenizer.process(entry.value);
      int index = _effectiveFields.indexOf(entry.key);
      if (index < 0 && fields.isEmpty) {
        _effectiveFields.add(entry.key);
        index = _effectiveFields.length - 1;
      }
      for (String token in tokens) {
        final indexedPropertyDocs = _data[token] ?? [];
        indexedPropertyDocs.add([_keys.length - 1, index]);
        _data[token] = indexedPropertyDocs;
      }
    }
  }

  void export(Sink<String> sink) {
    final encoder = JsonEncoder();
    final chunk = encoder.startChunkedConversion(sink);
    chunk.add({'keys': _keys, 'data': _data, 'fields': _effectiveFields});
    chunk.close();
  }

  void import(Map<String, dynamic> json) {
    _keys.addAll(json['keys']);
    _data.addAll(json['data']);
    _effectiveFields.addAll(json['fields']);
  }

  List<DocPropIndex>? find(String word) {
    return _data[word];
  }

  String keyOf(int docIndex) {
    return _keys[docIndex];
  }

  String fieldOf(int fieldIndex) {
    return _effectiveFields[fieldIndex];
  }
}

File: js/src/index_map.ts
import _ from "lodash";
import Tokenizer from "./tokenizer";

export type DocPropIndex = number[];

class IndexMap {
  private _data: Record<string, DocPropIndex[]> = {};
  private _keys: string[] = [];
  private _effectiveFields: string[] = [];
  private readonly tokenizer: Tokenizer;

  constructor({ tokenizer, fields = [] }: { tokenizer: Tokenizer; fields?: string[] }) {
    this.tokenizer = tokenizer;

    if (fields.length > 0) {
      this._effectiveFields.push(...fields);
    }
  }

  add(key: string, document: Record<string, any>): void {
    this._keys.push(key);

    _.forEach(document, (value, key) => {
        if (typeof value !== 'string' || (this._effectiveFields.length > 0 && !_.includes(this._effectiveFields, key))) {
        return;
      }
      const tokens = this.tokenizer.process(value);
      let index = this._effectiveFields.indexOf(key);

      if (index < 0 && this._effectiveFields.length === 0) {
        this._effectiveFields.push(key);
        index = this._effectiveFields.length - 1;
      }
      _.forEach(tokens, (token) => {
        const indexedPropertyDocs = _.get(this._data, token, [] as Array<DocPropIndex>);
        indexedPropertyDocs.push([this._keys.length - 1, index]);
        this._data[token] = indexedPropertyDocs;
      });
    });
  }

  export(): string {
    return JSON.stringify({
        keys: this._keys, data: this._data, fields: this._effectiveFields
    });
  }

  import(json: { keys: string[]; data: Record<string, DocPropIndex[]>; fields: string[] }): void {
    this._keys.push(...json.keys);
    Object.assign(this._data, json.data);
    this._effectiveFields.push(...json.fields);
  }

  find(word: string): DocPropIndex[] | undefined {
    return this._data[word];
  }

  keyOf(docIndex: number): string {
    return this._keys[docIndex];
  }

  fieldOf(fieldIndex: number): string {
    return this._effectiveFields[fieldIndex];
  }
}

export default IndexMap;